include/fiction/algorithms/simulation/sidb/critical_temperature.hpp

```diff
@@ -446,14 +386,71 @@ class critical_temperature_impl
      * SiDB cell-level layout.
      */
     Lyt layout{};
+
     /**
      * Parameters for the `critical_temperature` algorithm.
      */
-    const critical_temperature_params& parameter;
+    const critical_temperature_params<TT>& parameter;
     /**
      * Statistics.
      */
     critical_temperature_stats<Lyt>& temperature_stats;
+    /**
+     * Iterator that iterates over all possible input states.
+     */
+    bdl_input_iterator<Lyt> bii{};
+    /**
+     * Collects expected output values for a given input configuration.
+     *
+     * This function determines the expected output values for a specific input configuration
+     * based on the provided truth table and stores them in the `expected_outputs` vector.
+     * The function takes into account the number of variables and bits in the truth table
+     * to correctly extract the expected output values for different gate types.
+     *
+     * @param input_bit The index of the current input configuration.
+     * @return Vector representing the expected output for the given input.
+     */
+    [[nodiscard]] std::vector<bool> outputs_for_given_input(const uint64_t input_bit)
+    {
+        std::vector<bool> expected_outputs;
+        expected_outputs.reserve(parameter.truth_table.size());
+
+        std::transform(parameter.truth_table.begin(), parameter.truth_table.end(), std::back_inserter(expected_outputs),
+                       [input_bit](const auto& output) { return kitty::get_bit(output, input_bit) != 0u; });
+
+        return expected_outputs;
+    }
```

Comment:
This information is in the truth table already. Why is it important to copy it into another container?

Reply:
True. I thought it would make readability easier, but in this case it is more like code duplication.